### PR TITLE
4756: Use Adgangsplatformen authentication client when logging out

### PR DESCRIPTION
**Pull request by kasperg**
#### Link to issue

https://platform.dandigbib.org/issues/4756

#### Description

We currently use a separate secondary Adgangsplaformen client for
retrieving tokens without access to sensitive information. These 
tokens are suitable for use in client side applications.

However the secondary client is potentially shareable between 
installations and thus cannot be configured with a single valid 
logout url. This means that Adgangsplatform cannot redirect to the
correct site.

To address this we instead revoke the access token manually and then
generare a unique access token for an anonymous user using the
primary client.

We can then pass this token to Adgangsplatformen when logging out and
it will allow redirect to the corresponding site.

#### Checklist

- [x] My complies with [our coding guidelines](../docs/code_guidelines.md).
- [x] My code passes our static analysis suite. If not then I have added a comment explaining why this change should be exempt from the code standards and process.
- [x] My code passes our continuous integration process. If not then I have added a comment explaining why this change should be exempt from the code standards and process.

### CODE REVIEW

File: modules/ding_adgangsplatformen/ding_adgangsplatformen.module [hunk 1]
```diff
@@ -521,6 +547,34 @@ function ding_adgangsplatformen_has_token_client() {
...
+ *   The OAuth2 provider to use when revoking the token.
+ * @param string $access_token
+ *   The access token to revoke.
+ *
+ * @return bool
+ *   Whether revoking the token succeeded or not.
+ */
+function ding_adgangsplatformen_revoke_token(GenericProvider $provider, $access_token) {
+  $request = $provider->getAuthenticatedRequest('DELETE',
+    'https://login.bib.dk/oauth/revoke', $access_token);
```
Comment (cableman) [inline modules/ding_adgangsplatformen/ding_adgangsplatformen.module]: Not super happy about the URL placement here. All the other URL's are kept together in the `ding_adgangsplatformen_get_configuration()`.

Comment (cableman) [reply modules/ding_adgangsplatformen/ding_adgangsplatformen.module]: Maybe a minor rewrite of `ding_adgangsplatformen_get_configuration()` is requires as I remember that the return value is pasted directly to the oAuth library, which is not super happy about changes to the config array.

File: modules/ding_adgangsplatformen/ding_adgangsplatformen.module [hunk 2]
```diff
@@ -275,20 +275,50 @@ function ding_adgangsplatformen_login() {
...
+    try {
+      // Retrieve a new anonymous token and use that for the logout procedure. By
+      // using the authentication provider we ensure that Adgangsplatformen will
+      // redirect to the right site.
+      $auth_provider = ding_adgangsplatformen_get_provider(TRUE);
+      $token = $auth_provider->getAccessToken('password', [
+        // Using @ as username and password works as an anonymous user.
+        'username' => '@',
+        'password' => '@',
+      ]);
```
Comment (cableman) [inline modules/ding_adgangsplatformen/ding_adgangsplatformen.module]: I not sure here, but from usage of the open platform... should the `ting_agency` have been sent with the request here?

Comment (cableman) [reply modules/ding_adgangsplatformen/ding_adgangsplatformen.module]: In the form `@775100` for both username and password? It's how we do it af the open platform at authentication to get an anonymous token for a given library?

Comment (kasperg) [reply modules/ding_adgangsplatformen/ding_adgangsplatformen.module]: Not as I understand the situation. According to @vibjerg the important part here is the client id as that it is the client that knows which urls are allowed for logout redirects.

File: modules/ding_adgangsplatformen/ding_adgangsplatformen.module (hunk 1 above)
Comment (kasperg) [reply modules/ding_adgangsplatformen/ding_adgangsplatformen.module]: > All the other URL's are kept together in the ding_adgangsplatformen_get_configuration()

Apart from https://login.bib.dk/oauth/revoke in the current version, but OK - I will bite and see what can be done.

As I understand it [the OAuth library is looking for specific keys in the configuration](https://github.com/thephpleague/oauth2-client/blob/2.3.0/src/Provider/AbstractProvider.php#L95-L108). As long as we do not override these we should be fine. We already put in a lot of other options which are irrelevant to the provider.

**Pull request merged**